compiler/rustc_hir_typeck/src/delegation.rs

```diff
@@ -0,0 +1,374 @@
+            _ => {}
+        }
+
+        ty.super_visit_with(self)
+    }
+}
+
+struct DelegationArgFolder<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    info: &'a GenericDefsMap<'tcx>,
+}
+
+impl<'tcx, 'a> DelegationArgFolder<'tcx, 'a> {
+    fn new(tcx: TyCtxt<'tcx>, info: &'a GenericDefsMap<'tcx>) -> DelegationArgFolder<'tcx, 'a> {
+        DelegationArgFolder { tcx, info }
+    }
+}
+
+impl<'tcx, 'a> TypeFolder<TyCtxt<'tcx>> for DelegationArgFolder<'tcx, 'a> {
+    fn interner(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        if let ty::Infer(inf_ty) = ty.kind()
+            && let ty::InferTy::TyVar(vid) = inf_ty
+        {
+            let param = self.info.ty_defs[vid].clone();
+            let index = vid.as_u32() + self.info.region_defs.len() as u32;
+            return Ty::new_param(self.tcx, index, param.name);
+        };
+        ty.super_fold_with(self)
+    }
+
+    fn fold_region(&mut self, r: ty::Region<'tcx>) -> ty::Region<'tcx> {
+        let ty::RegionKind::ReVar(rid) = &r.kind() else {
+            return r;
+        };
+        let param = &self.info.region_defs[rid];
+        ty::Region::new_early_param(
+            self.tcx,
+            ty::EarlyParamRegion { index: rid.as_u32(), name: param.name },
+        )
+    }
+}
+
+struct DelegationLowerer<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    info: &'a GenericDefsMap<'tcx>,
+    def_id: LocalDefId,
+    sig_id: DefId,
+}
+
+impl<'tcx, 'a> DelegationLowerer<'tcx, 'a> {
+    fn new(
+        tcx: TyCtxt<'tcx>,
+        def_id: LocalDefId,
+        info: &'a GenericDefsMap<'tcx>,
+    ) -> DelegationLowerer<'tcx, 'a> {
+        DelegationLowerer { tcx, info, def_id, sig_id: tcx.hir().delegation_sig_id(def_id) }
+    }
+
+    fn fn_sig(&self) -> (&'tcx [Ty<'tcx>], Ty<'tcx>) {
+        let caller_sig = self.tcx.fn_sig(self.sig_id);
+
+        let caller_kind = fn_kind(self.tcx, self.def_id.into());
+        let callee_kind = fn_kind(self.tcx, self.sig_id);
+
+        let sig = match (caller_kind, callee_kind) {
+            (FnKind::Free, _) => {
+                let mut args = self.info.defs[&self.sig_id];
+                let mut folder = DelegationArgFolder::new(self.tcx, self.info);
+                args = args.fold_with(&mut folder);
+
+                caller_sig.instantiate(self.tcx, args)
+            }
+            // only `Self` param supported here
+            (FnKind::AssocTraitImpl, FnKind::AssocTrait)
+            | (FnKind::AssocInherentImpl, FnKind::AssocTrait) => {
+                let parent = self.tcx.parent(self.def_id.into());
+                let self_ty = self.tcx.type_of(parent).instantiate_identity();
+                let generic_self_ty = ty::GenericArg::from(self_ty);
+                let args = self.tcx.mk_args_from_iter(std::iter::once(generic_self_ty));
+                caller_sig.instantiate(self.tcx, args)
+            }
+            // here generics are not yet supported
```

Comment:
error instead?

Reply:
Error was reported earlier in check_constraints. I added a comment.